File: itx-java/itx-impl/src/main/java/io/github/chehsunliu/itx/impl/queue/rabbitmq/RabbitMessageQueueFactory.java
package io.github.chehsunliu.itx.impl.queue.rabbitmq;

import static io.github.chehsunliu.itx.impl.Env.envInt;
import static io.github.chehsunliu.itx.impl.Env.requireEnv;

import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import io.github.chehsunliu.itx.contract.queue.MessageQueue;
import io.github.chehsunliu.itx.contract.queue.MessageQueueFactory;
import io.github.chehsunliu.itx.contract.queue.QueueException;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public final class RabbitMessageQueueFactory implements MessageQueueFactory {
    private final Connection connection;
    private final int maxConcurrency;
    private final String controlStandardQueue;
    private final String controlPremiumQueue;
    private final String computeStandardQueue;
    private final String computePremiumQueue;

    public static RabbitMessageQueueFactory fromEnv() {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(requireEnv("ITX_RABBITMQ_HOST"));
        factory.setPort(Integer.parseInt(requireEnv("ITX_RABBITMQ_PORT")));
        factory.setUsername(requireEnv("ITX_RABBITMQ_USER"));
        factory.setPassword(requireEnv("ITX_RABBITMQ_PASSWORD"));
        Connection conn;
        try {
            conn = factory.newConnection("itx-java");
        } catch (Exception e) {
            throw new QueueException("failed to open RabbitMQ connection", e);
        }
        return new RabbitMessageQueueFactory(
                conn,
                envInt("ITX_RABBITMQ_MAX_CONCURRENCY", 100),
                requireEnv("ITX_RABBITMQ_CONTROL_STANDARD_QUEUE"),
                requireEnv("ITX_RABBITMQ_CONTROL_PREMIUM_QUEUE"),
                requireEnv("ITX_RABBITMQ_COMPUTE_STANDARD_QUEUE"),
                requireEnv("ITX_RABBITMQ_COMPUTE_PREMIUM_QUEUE"));
    }

    @Override
    public MessageQueue createControlStandardQueue() {
        return new RabbitMessageQueue(connection, controlStandardQueue, maxConcurrency);
    }

    @Override
    public MessageQueue createControlPremiumQueue() {
        return new RabbitMessageQueue(connection, controlPremiumQueue, maxConcurrency);
    }

    @Override
    public MessageQueue createComputeStandardQueue() {
        return new RabbitMessageQueue(connection, computeStandardQueue, maxConcurrency);
    }

    @Override
    public MessageQueue createComputePremiumQueue() {
        return new RabbitMessageQueue(connection, computePremiumQueue, maxConcurrency);
    }
}

File: itx-kt/itx-impl/src/main/kotlin/io/github/chehsunliu/itx/impl/queue/rabbitmq/RabbitMessageQueueFactory.kt
package io.github.chehsunliu.itx.impl.queue.rabbitmq

import com.rabbitmq.client.Connection
import com.rabbitmq.client.ConnectionFactory
import io.github.chehsunliu.itx.contract.queue.MessageQueue
import io.github.chehsunliu.itx.contract.queue.MessageQueueFactory
import io.github.chehsunliu.itx.contract.queue.QueueException
import io.github.chehsunliu.itx.impl.envInt
import io.github.chehsunliu.itx.impl.requireEnv

class RabbitMessageQueueFactory(
    private val connection: Connection,
    private val maxConcurrency: Int,
    private val controlStandardQueue: String,
    private val controlPremiumQueue: String,
    private val computeStandardQueue: String,
    private val computePremiumQueue: String,
) : MessageQueueFactory {
    companion object {
        fun fromEnv(): RabbitMessageQueueFactory {
            val factory =
                ConnectionFactory().apply {
                    host = requireEnv("ITX_RABBITMQ_HOST")
                    port = requireEnv("ITX_RABBITMQ_PORT").toInt()
                    username = requireEnv("ITX_RABBITMQ_USER")
                    password = requireEnv("ITX_RABBITMQ_PASSWORD")
                }
            val conn =
                try {
                    factory.newConnection("itx-kt")
                } catch (e: Exception) {
                    throw QueueException("failed to open RabbitMQ connection", e)
                }
            return RabbitMessageQueueFactory(
                connection = conn,
                maxConcurrency = envInt("ITX_RABBITMQ_MAX_CONCURRENCY", 100),
                controlStandardQueue = requireEnv("ITX_RABBITMQ_CONTROL_STANDARD_QUEUE"),
                controlPremiumQueue = requireEnv("ITX_RABBITMQ_CONTROL_PREMIUM_QUEUE"),
                computeStandardQueue = requireEnv("ITX_RABBITMQ_COMPUTE_STANDARD_QUEUE"),
                computePremiumQueue = requireEnv("ITX_RABBITMQ_COMPUTE_PREMIUM_QUEUE"),
            )
        }
    }

    override fun createControlStandardQueue(): MessageQueue = RabbitMessageQueue(connection, controlStandardQueue, maxConcurrency)

    override fun createControlPremiumQueue(): MessageQueue = RabbitMessageQueue(connection, controlPremiumQueue, maxConcurrency)

    override fun createComputeStandardQueue(): MessageQueue = RabbitMessageQueue(connection, computeStandardQueue, maxConcurrency)

    override fun createComputePremiumQueue(): MessageQueue = RabbitMessageQueue(connection, computePremiumQueue, maxConcurrency)
}
